benchmark/scenarios/_request_load.py

```diff
@@ -0,0 +1,234 @@
+    :ivar timeout: Maximum time in seconds to wait for the requested
+        rate to be reached.
+    :ivar maintained: A `Deferred` that fires with an errback if the desired
+        scenario fails to hold between being established and being
+        stopped.  This Deferred never fires with a callback.
+    :ivar rate_measurer: `RateMeasurer` instace to monitor the scenario.
+    :ivar loop: main loop that will be doing requests every second.
+    :ivar monitor_loop: loop that will monitor the status of the scenario once
+        the target has been reached.
+    """
+
+    def __init__(
+        self, reactor, scenario_setup_instance, request_rate=10,
+        sample_size=DEFAULT_SAMPLE_SIZE, timeout=45
+    ):
+        self.reactor = reactor
+        self.scenario_setup = scenario_setup_instance
+        self.request_rate = request_rate
+        self.timeout = timeout
+        self._maintained = Deferred()
+        self.rate_measurer = RateMeasurer(sample_size)
+        self.max_outstanding = 10 * request_rate
+        # Send requests per second
+        self.loop = LoopingCall.withCount(self._request_and_measure)
+        self.loop.clock = self.reactor
+        # Monitor the status of the scenario
+        self.monitor_loop = LoopingCall(self.check_rate)
+        self.monitor_loop.clock = self.reactor
+
+    def _request_and_measure(self, count):
+        """
+        Update the rate with the current value and send `request_rate`
+        number of new requests.
+
+        :param count: The number of seconds passed since the last time
+            `_request_and_measure` was called.
+        """
+        for i in range(count):
+            self.rate_measurer.update_rate()
+
+        def handle_request_error(result):
+            self.rate_measurer.request_failed(result)
+            write_failure(result)
+
+        for i in range(self.request_rate):
+            d = self.scenario_setup.make_request()
+            self.rate_measurer.request_sent()
+            d.addCallbacks(self.rate_measurer.response_received,
+                           errback=handle_request_error)
+
+    def _fail(self, exception):
+        """
+        Fail the scenario. Stop the monitor loop and throw the
+        error.
+
+        :param exception: `Exception` that caused the failure.
+        """
+        self.monitor_loop.stop()
+        self._maintained.errback(exception)
+
+    def check_rate(self):
+        """
+        Verify that the rate hasn't decreased and that the scenario is
+        not overloaded. A scenario is overloaded if there are too many
+        outstanding requests.
+
+        :raise RequestRateTooLow: if the rate has dropped.
+        :raise RequestOverload: if the scenario is overloaded.
+        """
+        rate = self.rate_measurer.rate()
+        if rate < self.request_rate:
+            self._fail(RequestRateTooLow(rate))
+
+        if self.rate_measurer.outstanding() > self.max_outstanding:
+            self._fail(RequestOverload())
```

Comment:
Both of the `_fail` calls in this method could run on a single call to `check_rate`.  This would double call  `self._maintained.errback` - which is an error and will raise `AlreadyCalledError`.